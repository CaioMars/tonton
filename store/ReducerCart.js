import { createSlice } from "@reduxjs/toolkit";

export const cartSlice = createSlice({
  name: "cart",
  initialState: {
    items: [],
    subtotal: 0,
    delivery: 0,
    discount: 0,
    total: 0,
  },
  reducers: {
    add: (state, action) => {
      const { payload } = action;
      state.subtotal += payload.price * payload.quantity;
      state.total = state.subtotal + state.delivery - state.discount;
      state.items.push(payload);
    },
    remove: (state, action) => {
      const item = state.items[action.payload];
      if (typeof item === "undefined") return;
      state.subtotal -= item.price * item.quantity;
      state.total = state.subtotal + state.delivery - state.discount;
      state.items = state.items.filter(
        (item, index) => index !== action.payload
      );
    },
  },
});

export const { add, remove } = cartSlice.actions;

export default cartSlice.reducer;
