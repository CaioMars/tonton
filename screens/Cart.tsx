import { FlatList, StyleSheet, Image, TouchableOpacity } from "react-native";
import { Text, View } from "../components/Themed";
import { Button } from "react-native-elements";
import { FontAwesome } from "@expo/vector-icons";
import { useDispatch, connect } from "react-redux";
import { remove } from "../store/ReducerCart";
import { twoDecimals } from "../functions/helper";
import Toast from "react-native-toast-message";

function Cart({ navigation, cart }: any) {
  const items = cart.items;
  const dispatch = useDispatch();

  if (items.length === 0) {
    return (
      <View style={styles.container}>
        <Image
          source={require("../assets/images/empty-cart.png")}
          resizeMode="contain"
          style={styles.ImageEmptyCart}
        />
        <Text style={styles.title}>Cart</Text>
        <Text style={{ marginBottom: 20 }}>Your cart is empty.</Text>
        <Button title="Go shopping" onPress={() => navigation.goBack()} />
      </View>
    );
  } else {
    return (
      <FlatList
        data={items}
        keyExtractor={(item: object, index: any) => index}
        renderItem={({ item, index }: any) => (
          <View key={index} style={styles.containerItem}>
            <View style={styles.containerImage}>
              <Image
                source={{ uri: item.image }}
                style={styles.image}
                resizeMode="contain"
              />
            </View>
            <View style={styles.containerTitle}>
              <Text>{item.title}</Text>
            </View>
            <View style={styles.containerPrice}>
              <Text>$ {twoDecimals(item.price)}</Text>
              <Text>QT: {item.quantity}</Text>
            </View>
            <View style={styles.containerDelete}>
              <TouchableOpacity onPress={() => dispatch(remove(index))}>
                <FontAwesome name="trash" size={20} />
              </TouchableOpacity>
            </View>
          </View>
        )}
        ListFooterComponent={() => {
          return (
            <>
              <View style={styles.containerTotal}>
                <View style={styles.containerTotalRow}>
                  <View style={styles.containerTotalLabel}>
                    <Text style={styles.textLabel}>Subtotal</Text>
                  </View>
                  <View style={styles.containerTotalValue}>
                    <Text style={styles.textValue}>
                      $ {twoDecimals(cart.subtotal)}
                    </Text>
                  </View>
                </View>
                <View style={styles.containerTotalRow}>
                  <View style={styles.containerTotalLabel}>
                    <Text style={styles.textLabel}>Delivery</Text>
                  </View>
                  <View style={styles.containerTotalValue}>
                    <Text style={styles.textValue}>
                      $ {twoDecimals(cart.delivery)}
                    </Text>
                  </View>
                </View>
                <View style={styles.containerTotalRow}>
                  <View style={styles.containerTotalLabel}>
                    <Text style={styles.textLabel}>Discount</Text>
                  </View>
                  <View style={styles.containerTotalValue}>
                    <Text style={styles.textValue}>
                      $ {twoDecimals(cart.discount)}
                    </Text>
                  </View>
                </View>
                <View style={styles.containerTotalRow}>
                  <View style={styles.containerTotalLabel}>
                    <Text style={styles.textTotal}>Total</Text>
                  </View>
                  <View style={styles.containerTotalValue}>
                    <Text style={styles.textTotal}>
                      ${twoDecimals(cart.total)}
                    </Text>
                  </View>
                </View>
              </View>
              <View style={styles.containerShopping}>
                <Button
                  title="Continue shopping"
                  onPress={() => navigation.goBack()}
                />
              </View>
              <View style={styles.containerCheckout}>
                <Button
                  title="Checkout"
                  type="outline"
                  onPress={() => {
                    Toast.show({
                      type: "success",
                      text1: "ℹ️   Feature beyond the scope of this project",
                      position: "bottom",
                    });
                  }}
                />
              </View>
            </>
          );
        }}
      />
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  title: {
    fontSize: 20,
    fontWeight: "bold",
  },
  containerItem: {
    flexDirection: "row",
    padding: 10,
    marginBottom: 5,
  },
  containerImage: {},
  containerTitle: {
    flex: 1,
    paddingHorizontal: 10,
    justifyContent: "center",
    alignItems: "center",
  },
  containerPrice: {
    paddingHorizontal: 20,
    justifyContent: "center",
    alignItems: "center",
  },
  containerDelete: {
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: 10,
  },
  image: {
    width: 75,
    height: 75,
  },
  containerTotal: {
    display: "flex",
    flexDirection: "column",
    flex: 1,
    marginBottom: 20,
    paddingHorizontal: 10,
    paddingVertical: 10,
  },
  containerTotalRow: {
    flexDirection: "row",
    padding: 5,
  },
  containerTotalLabel: {
    flex: 1,
  },
  containerTotalValue: {
    flex: 1,
    alignItems: "flex-end",
  },
  containerShopping: {
    padding: 10,
  },
  containerCheckout: {
    padding: 10,
  },
  textLabel: {},
  textValue: {},
  textTotal: {
    fontWeight: "bold",
    fontSize: 18,
  },
  ImageEmptyCart: {
    width: "100%",
    maxWidth: 495,
  },
});

function mapStateToProps(state: any) {
  return { cart: state.cart };
}

export default connect(mapStateToProps)(Cart);
