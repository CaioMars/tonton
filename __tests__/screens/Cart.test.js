jest.useFakeTimers();
import ReducerCart, { add } from "../../store/ReducerCart";
import Cart from "../../screens/Cart";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { configureStore } from "@reduxjs/toolkit";

describe("<Cart />", () => {
  let cart, store, tree;

  it("cart with NO items renders correctly", async () => {
    store = configureStore({
      reducer: {
        cart: ReducerCart,
      },
    });

    tree = await renderer
      .create(
        <Provider store={store}>
          <Cart cart={store.getState().cart} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  const testItem1 = {
    id: 1,
    title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
    price: 109.95,
    description:
      "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
    category: "men's clothing",
    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    rating: { rate: 3.9, count: 120 },
  };

  it("cart with items renders correctly", async () => {
    cart = ReducerCart(
      { items: [], subtotal: 0, delivery: 0, discount: 0, total: 0 },
      add(testItem1)
    );
    const preloadedState = { cart };
    store = configureStore({
      reducer: {
        cart: ReducerCart,
      },
      preloadedState,
    });

    tree = await renderer
      .create(
        <Provider store={store}>
          <Cart cart={cart} />
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
