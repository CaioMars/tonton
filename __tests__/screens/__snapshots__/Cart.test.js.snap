// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> cart with NO items renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <Image
    resizeMode="contain"
    source={1}
    style={
      Object {
        "maxWidth": 495,
        "width": "100%",
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        },
      ]
    }
  >
    Cart
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
        },
        Object {
          "marginBottom": 20,
        },
      ]
    }
  >
    Your cart is empty.
  </Text>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Go shopping
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Cart /> cart with items renders correctly 1`] = `
<RCTScrollView
  ListFooterComponent={[Function]}
  data={
    Array [
      Object {
        "category": "men's clothing",
        "description": "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        "id": 1,
        "image": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        "price": 109.95,
        "rating": Object {
          "count": 120,
          "rate": 3.9,
        },
        "title": "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "flexDirection": "row",
              "marginBottom": 5,
              "padding": 10,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {},
            ]
          }
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "uri": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
              }
            }
            style={
              Object {
                "height": 75,
                "width": 75,
              }
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
                "paddingHorizontal": 10,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                },
                undefined,
              ]
            }
          >
            Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingHorizontal": 20,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                },
                undefined,
              ]
            }
          >
            $ 
            109.95
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "paddingHorizontal": 10,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
              "marginBottom": 20,
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "flexDirection": "row",
                "padding": 5,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {},
                ]
              }
            >
              Subtotal
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "alignItems": "flex-end",
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {},
                ]
              }
            >
              $ 
              109.95
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "flexDirection": "row",
                "padding": 5,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {},
                ]
              }
            >
              Delivery
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "alignItems": "flex-end",
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {},
                ]
              }
            >
              $ 
              0.00
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "flexDirection": "row",
                "padding": 5,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {},
                ]
              }
            >
              Discount
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "alignItems": "flex-end",
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {},
                ]
              }
            >
              $ 
              0.00
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#fff",
              },
              Object {
                "flexDirection": "row",
                "padding": 5,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              Total
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#fff",
                },
                Object {
                  "alignItems": "flex-end",
                  "flex": 1,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000",
                  },
                  Object {
                    "fontSize": 18,
                    "fontWeight": "bold",
                  },
                ]
              }
            >
              $
              109.95
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "padding": 10,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 3,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 3,
              },
              undefined,
              false,
            ]
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "busy": false,
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#2089dc",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "white",
                    "fontSize": 18,
                    "paddingVertical": 1,
                    "textAlign": "center",
                  }
                }
              >
                Continue shopping
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "padding": 10,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 3,
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 3,
              },
              undefined,
              false,
            ]
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "busy": false,
                "disabled": false,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "borderColor": "#2089dc",
                  "borderRadius": 3,
                  "borderWidth": 0.5,
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2089dc",
                    "fontSize": 18,
                    "paddingVertical": 1,
                    "textAlign": "center",
                  }
                }
              >
                Checkout
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
